core/src/main/java/io/lionweb/lioncore/java/model/impl/DynamicNode.java

```diff
@@ -70,14 +71,40 @@ public boolean equals(Object o) {
     }
     DynamicNode that = (DynamicNode) o;
     return Objects.equals(id, that.id)
-        && Objects.equals(parent, that.parent)
-        && Objects.equals(concept, that.concept)
+        && shallowNodeEquality(parent, that.parent)
+        && shallowNodeEquality(concept, that.concept)
         && Objects.equals(propertyValues, that.propertyValues)
-        && Objects.equals(containmentValues, that.containmentValues)
+        && shallowContainmentsEquality(containmentValues, that.containmentValues)
         && Objects.equals(referenceValues, that.referenceValues)
         && Objects.equals(annotations, that.annotations);
   }
 
+  private static boolean shallowContainmentsEquality(
+      Map<String, List<Node>> containments1, Map<String, List<Node>> containments2) {
+    if (!containments1.keySet().equals(containments2.keySet())) {
+      return false;
+    }
+    return containments1.keySet().stream()
+        .allMatch(
+            containmentName -> {
+              List<Node> nodes1 = containments1.get(containmentName);
+              List<Node> nodes2 = containments2.get(containmentName);
+              return nodes1.size() == nodes2.size()
+                  && IntStream.range(0, nodes1.size())
+                      .allMatch(i -> shallowNodeEquality(nodes1.get(i), nodes2.get(i)));
+            });
+  }
+
+  private static boolean shallowNodeEquality(@Nullable Node node1, @Nullable Node node2) {
+    if (node1 == null || node2 == null) {
+      return true;
```

Comment:
Should only return `true` when both are `null`, right?

Reply:
Right!

Reply:
But if exactly one of `node1`, `node2` is `null`, then this method falls through to `Objects.equals(node1, node2)` again. Functionally that's the same (I guess), but we could add handling this case easily here, and would improve understandability IMHO.